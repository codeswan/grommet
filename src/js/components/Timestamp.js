// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP

import React, { Component, PropTypes } from 'react';
import { getCurrentLocale } from '../utils/Locale';
import CSSClassnames from '../utils/CSSClassnames';

const CLASS_ROOT = CSSClassnames.TIMESTAMP;

function _showField(field, fields) {
  let result = true;
  if (fields) {
    if (Array.isArray(fields)) {
      result = fields.indexOf(field) !== -1;
    } else {
      result = field === fields;
    }
  }
  return result;
}

export default class Timestamp extends Component {

  constructor(props, context) {
    super(props, context);
    this.state = {};
  }

  componentDidMount () {
    this._formatForLocale(this.props);
  }

  componentWillReceiveProps (nextProps) {
    this._formatForLocale(nextProps);
  }

  _formatForLocale (props) {
    const locale = getCurrentLocale();
    const value =
      (typeof props.value === 'string') ? new Date(props.value) : props.value;

    let date;
    if (_showField('date', props.fields)) {
      const dateOptions = { year: 'numeric', month: 'short', day: 'numeric' };
      date = value.toLocaleDateString(locale, dateOptions);
    }

    let time;
    if (_showField('time', props.fields)) {
      const timeOptions = { hour: '2-digit', minute: '2-digit' };
      time = value.toLocaleTimeString(locale, timeOptions);
    }

    this.setState({ date: date, time: time });
  }

  render () {
    var classes = [CLASS_ROOT];
    classes.push(CLASS_ROOT + '--' + this.props.align);
    if (this.props.className) {
      classes.push(this.props.className);
    }

    let date;
    if (this.state.date) {
      date = <span className={`${CLASS_ROOT}__date`}>{this.state.date}</span>;
    }

    let time;
    if (this.state.time) {
      time = <span className={`${CLASS_ROOT}__time`}>{this.state.time}</span>;
    }

    return (
      <span className={classes.join(' ')}>
        {date} {time}
      </span>
    );
  }

}

const FIELD_TYPES = PropTypes.oneOf(['date', 'time']);

Timestamp.propTypes = {
  align: PropTypes.oneOf(['left', 'right']),
  fields: PropTypes.oneOfType([
    PropTypes.arrayOf(FIELD_TYPES),
    FIELD_TYPES
  ]),
  value: PropTypes.oneOfType([
    PropTypes.string, // ISO-8601 string
    PropTypes.object  // Date object
  ]).isRequired
};

Timestamp.defaultProps = {
  align: 'left'
};
