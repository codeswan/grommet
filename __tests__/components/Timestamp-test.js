// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP

import React from 'react';
import renderer from 'react/lib/ReactTestRenderer';

import Timestamp from '../../src/js/components/Timestamp';

describe('Timestamp', () => {
  it('has correct default options', () => {
    const component = renderer.create(
      <Timestamp value='4/7/2015 10:00 am' />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
