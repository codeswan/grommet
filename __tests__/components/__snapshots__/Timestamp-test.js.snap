exports[`Timestamp has correct default options 1`] = `
<span
  className="grommetux-timestamp grommetux-timestamp--left">
  <span
    className="grommetux-timestamp__date">
    Apr 7, 2015
  </span>
   
  <span
    className="grommetux-timestamp__time">
    10:00 AM
  </span>
</span>
`;
